extensions-core/druid-catalog/src/main/java/org/apache/druid/catalog/http/CatalogResource.java

```diff
@@ -0,0 +1,586 @@
+import org.apache.druid.server.security.Action;
+import org.apache.druid.server.security.AuthorizationUtils;
+import org.apache.druid.server.security.AuthorizerMapper;
+import org.apache.druid.server.security.ForbiddenException;
+import org.apache.druid.server.security.Resource;
+import org.apache.druid.server.security.ResourceAction;
+import org.apache.druid.server.security.ResourceType;
+
+import javax.inject.Inject;
+import javax.servlet.http.HttpServletRequest;
+import javax.ws.rs.Consumes;
+import javax.ws.rs.DELETE;
+import javax.ws.rs.GET;
+import javax.ws.rs.POST;
+import javax.ws.rs.Path;
+import javax.ws.rs.PathParam;
+import javax.ws.rs.Produces;
+import javax.ws.rs.QueryParam;
+import javax.ws.rs.core.Context;
+import javax.ws.rs.core.MediaType;
+import javax.ws.rs.core.Response;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.stream.Collectors;
+
+/**
+ * REST endpoint for user and internal catalog actions. Catalog actions
+ * are divided by operation: configuration-as-code, edits, retrieval,
+ * etc. Operations occur at the global level (all schemas), the schema level, or the
+ * table level.
+ *
+ * @see {@link CatalogListenerResource} for the broker-side API.
+ */
+@Path(CatalogResource.ROOT_PATH)
+public class CatalogResource
+{
+  public static final String ROOT_PATH = "/druid/coordinator/v1/catalog";
+
+  public static final String NAME_FORMAT = "name";
+  public static final String PATH_FORMAT = "path";
+  public static final String METADATA_FORMAT = "metadata";
+  public static final String STATUS_FORMAT = "status";
+
+  private final CatalogStorage catalog;
+  private final AuthorizerMapper authorizerMapper;
+
+  @Inject
+  public CatalogResource(
+      final CatalogStorage catalog,
+      final AuthorizerMapper authorizerMapper
+  )
+  {
+    this.catalog = catalog;
+    this.authorizerMapper = authorizerMapper;
+  }
+
+  // ---------------------------------------------------------------------
+  // Configuration-as-code style methods
+
+  /**
+   * Create or update a new table containing the given table specification.
+   * Supports three use cases:
+   * <ul>
+   * <li>"create if not exists": default use case with no options.</li>
+   * <li>"create or update": set {@code overwrite=true}.</li>
+   * <li>"update": set {@code version} to the expected current version.
+   *     This form enforces optimistic locking.</li>
+   * </ul>
+   *
+   * @param dbSchema The name of the Druid schema, which must be writable
+   *        and the user must have at least read access.
+   * @param name The name of the table definition to modify. The user must
+   *        have write access to the table.
+   * @param spec The new table definition.
+   * @param version the expected version of an existing table. The version must
+   *        match. If not (or if the table does not exist), returns an error.
+   * @param overwrite if {@code true}, then overwrites any existing table.
```

Comment:
Would be good to indicate that this property is ignored, if version is specified.